python/src/main/java/org/apache/apex/malhar/python/base/BasePythonExecutionOperator.java

```diff
@@ -0,0 +1,591 @@
+  }
+
+  @Override
+  public void deactivate()
+  {
+
+  }
+
+  /***
+   * Used to emit the stragglers into the Stragglers port. INvoked both when a new tuple arrives or when Idle time
+   *  is detected on this operator.
+   */
+  private void emitStragglers()
+  {
+    List<PythonRequestResponse> stragglerResponse = new ArrayList<>();
+    getApexPythonEngine().getDelayedResponseQueue().drainTo(stragglerResponse);
+    for (PythonRequestResponse aReqResponse : stragglerResponse) {
+      responseCounterForThisWindow += 1;
+      stragglersPort.emit(aReqResponse);
+    }
+  }
+
+  /***
+   * Instantiates the configured python engine. Only in-memory implementation provided for now
+   * @param context The operator context
+   * @return The python engine
+   */
+  protected ApexPythonEngine initApexPythonEngineImpl(Context.OperatorContext context)
+  {
+    JepPythonEngine jepPythonEngine = new JepPythonEngine("" + context.getId(),workerThreadPoolSize);
+    jepPythonEngine.setSleepTimeAfterInterpreterStart(sleepTimeDuringInterpreterBoot);
+    return jepPythonEngine;
+  }
+
+  /***
+   * Instantiates the partitioner. Only Thread Starvation based partitioner for now.
+   */
+  private void initPartitioner()
+  {
+    if (partitioner == null) {
+      synchronized (objectForLocking) {
+        if (partitioner == null) {
+          switch (partitionerType) {
+            default:
+            case THREAD_STARVATION_BASED:
+              partitioner = new ThreadStarvationBasedPartitioner(this);
+              break;
+          }
+        }
+      }
+    }
+  }
+
+  /***
+   * Starts the python engine and also sleeps for sometime ( configurable) to ensure that the interpreter is
+   *  completely booted up in memory. The time taken to boot the interpreter depends on the libaries that are loaded etc
+   * @param context The Operator context
+   */
+  @Override
+  public void setup(Context.OperatorContext context)
+  {
+    super.setup(context);
+    apexPythonEngine = initApexPythonEngineImpl(context);
+    Map<PythonInterpreterConfig,Object> preInitConfigurations = getPreInitConfigurations();
+    apexPythonEngine.preInitInterpreter(preInitConfigurations);
+    apexPythonEngine.startInterpreter();
+    apexPythonEngine.postStartInterpreter();
+    initPartitioner();
+    processPostSetUpPythonInstructions(apexPythonEngine);
+  }
+
+  @Override
+  public void teardown()
+  {
+    super.teardown();
+    apexPythonEngine.stopInterpreter();
+  }
+
+  @Override
+  public void beginWindow(long windowId)
+  {
+    super.beginWindow(windowId);
+    requestCounterForThisWindow = 0;
+    responseCounterForThisWindow = 0;
+    numNullResponsesPerWindow = 0;
+    currentWindowId = windowId;
+  }
+
+  @Override
+  public void endWindow()
+  {
+    super.endWindow();
+    if (responseCounterForThisWindow < requestCounterForThisWindow) {
+      LOG.info("Detected stragglers and configured flag/state for blocking at end of window is " +
```

Comment:
should be debug and possibly not logged repeatedly?

Reply:
Changed to debug and moved the other repeated log statement out of the loop.